hydroflow/examples/two_pc/coordinator.rs

```diff
@@ -0,0 +1,220 @@
+    // with the transaction ID and a unique message ID.
+    // HACK: We move the xids HashSet into a map operator as flow state.
+    // Should be done with a groupby.
+    let mut xids = HashSet::new();
+    hf.add_subgraph(
+        "phase 1 init",
+        text_out
+            .flatten()
+            .filter_map(move |xidstr| {
+                match xidstr.trim().parse::<u16>() {
+                    Ok(the_xid) => {
+                        if xids.contains(&the_xid) {
+                            println!("Transaction ID {} already used", the_xid);
+                            None
+                        } else {
+                            xids.insert(the_xid);
+                            Some(CoordMsg {
+                                xid: the_xid,
+                                mid: 1, // first message for this transaction
+                                mtype: MsgType::Prepare,
+                            })
+                        }
+                    }
+                    Err(_) => None,
+                }
+            })
+            .cross_join(subordinates_out_tee_1_pull.flatten())
+            .map(|(msg, addr)| Some((addr, msg)))
+            .pull_to_push()
+            .push_to(p1_send_push),
+    );
+
+    // Phase 1 Response and Phase 2 Request:
+    // Receive and count subordinate responses until we have them all.
+    // Once we have them all, assess the unanimous vote condition for commit and
+    // send either a Commit or Abort message.
+    // HACK: We move the vote_ctr((xid, mid), count) HashMap into a map operator as flow state.
+    // Should be done with a groupby.
+    let mut vote_ctr = HashMap::new();
+    hf.add_subgraph(
+        "collect votes and send command",
+        p1_recv_pull
+            .flatten()
+            .map(move |msg| {
+                println!("{:?}", msg);
+                msg
+            })
+            .filter_map(move |msg| {
+                *vote_ctr.entry((msg.xid, msg.mtype)).or_insert(0) += 1;
+                // HACK: the constant 3 here is the length of the subordinates list!
+                // Should be computed on bootstrap tick and joined in here.
+                if vote_ctr.get(&(msg.xid, MsgType::Commit)).unwrap_or(&0)
+                    + vote_ctr.get(&(msg.xid, MsgType::Abort)).unwrap_or(&0)
+                    == 3
+                {
+                    // Abort if any subordinate voted to Abort
+                    if vote_ctr.get(&(msg.xid, MsgType::Abort)).unwrap_or(&0) > &0 {
+                        Some(CoordMsg {
+                            xid: msg.xid,
+                            mid: msg.mid + 1,
+                            mtype: MsgType::Abort,
+                        })
+                    } else {
+                        Some(CoordMsg {
+                            xid: msg.xid,
+                            mid: msg.mid + 1,
+                            mtype: MsgType::Commit,
+                        })
+                    }
```

Comment:
Oh you can also rearrange this to reduce duplicated code:
```rust
Some(CoordMsg {
    xid: msg.xid,
    mid: msg.mid + 1,
    // Abort if any subordinate voted to Abort
    mtype: if vote_ctr.get(&(msg.xid, MsgType::Abort)).unwrap_or(&0) > &0 {
        MsgType::Abort,
    } else {
        MsgType::Commit,
    }
}
```

Reply:
nice!